### ISSUE TITLE
Report Reviewer 2

### ISSUE THREAD

**Issue by MarieEtienne** [computorg/published-202402-lefort-peerannot#5]
Associate Editor: *MP Etienne*
Reviewer : [name](url) (chose to lift his/her anonymity / remain anonymous)

## Reviewer: Reviewing history

- Paper submitted *September, 8 2023*
- Reviewer invited *September, 21 2023*
- Review 1 received *November, 17, 2023*
- Paper revised *November, 23, 2023*
- Reviewer invited *September, 21 2023*
- Review 2 received *December, 13, 2023*
- Paper conditionally accepted *December, 18, 2023*

### Summary Of Contributions:

This article depicts a new python package released by the authors to analyze crowdsourced annotations in the context of (image) classification. The article gives an overview of existing methods to handle such data, especially the way of aggregating the classifications given by the different workers (persons that annotates).

### Strengths And Weaknesses:

In my opinion, this work aims at fitting one of the scope of Computo, namely
 -   Software/tutorial papers to present implementations of stats/ML algorithms or to feature the use of a package/toolbox. For such papers we expect not only the description of an existing implementation but also the study of a concrete use case. If applicable, a comparison to related works and appropriate benchmarking are also expected. (Source: Computo website).

As a first remark, I would say that the study of a concrete use case is a bit light in my opinion, as the two datasets CIPHAR-H10 and LabelMe are clearly toy datasets which are barely described and whose purpose is not discuss. The authors discuss the difficulty of finding datasets in this domain, but if they could not manage to find other ones, it should be interesting to describe more the one used and their potential interest. Moreover, it is not discussed whether comparable related works exist for this task, and thus this tool is not compared to any other (but comparable tools might not exist). From this point of view, I am not in favor of an acceptation in the journal from its current form, even if I let the editors judge whether it is suited for Computo. From a reproductible point of view, the authors provide all working codes and all the results and figure depicted in the articles can be reproduced.

This being said, the article, in my opinion, sometimes lack of clarity in its writing and always oscillates between a description of the package and problematics of aggregation in crowdsourcing. I think it might be valuable to make all necessary reminders about the methods for crowdsourced datasets (which are necessary and, overall, well done), and then dive into the proposed package, with maybe a bit more precision about arguments of the main functions (see my comments below).

As a major confusion for me during my reading is the fact that the authors emphasize in the introduction that: "With crowdsourced data the ground truth of a task
, denoted

is unknown, and there is no single label that can be trusted as in standard supervised learning (even on the train set!)", and below that : "For both CIFAR-10H and LabelMe, the dataset was originally released in classical supervised learning form (without crowdsourcing). These labels are used as ground truth in evaluations and visualizations. However, we emphasize that crowdsourcing strategies do not rely on the ground truth (only on the workers' answers). However, from what I understood, all the analysis made using the package are done using the available ground truth, and thus it is not clear how this package can provide nice insights in the (interesting) setting where ground truth do not exist (for instance, how do we train a classification model, ho do we detect a bad worker, etc...). I think this should be greatly clarified by the authors for further submission (in this journal or another one).
### Changes And Questions:
 *Major remarks*

    In section 2.1, the authors well depict the required dataset format for the package, and especially the .json file has depicted in Figure 2. They say that it is more memory friendly than a full matrix where -1 would code missing values. However, I was wondering if the package propose a function to encode a .json file in the right format from such a matrix (which is probably a current way of doing such annotations in this field)?
    In section 3.1.3 the authors write the likelihood in the context of known labels, whereas they wrote before that the with crowdsourced data, the ground truth of a task [\dots] is unknown, thus I think the authors should focus on the likelihood in this case (case 2 of the original paper of Dawid and Skene's). From a didactic point of view, writing this likelihood and then explaining that strategies like EM algorithm are relevant for optimization (which is not necessary whan the ground truth is known) in this case seems to be in a better order for me.
    On section 3.2, the authors linger for a long time over simulation under different scenarios to (from what I understand) i) depicts their simulation function; ii) discuss in which scenario the different aggregation strategy should be appropriate. I understand that the authors show this feature of the package, but in my opinion, it would be more interesting to work with athe real data sets studied by the authors (as they have a ground truth), then to show the accuracy table on this data set, and to discuss, depending on the best aggregation strategy, what it tells us about the annotation scheme (it highlights that their might be cluster of worker, for instance...). I found the exercise of finding that the best model is the one from which simulations were made a bit long here.
    Section 4.1.1. The quantities 

    are the confusion matrices of each worker, but again, this imposes that the ground truth is known, isn't it? Again, section 4.2 poses on the case where ground truth is known. I find this a bit confusing as a reader, maybe the authors should clearly reorganize the paper, having a section that clearly covers the case of crowdsourcing with available ground truth and a one without available ground truth. As most of the sections rely at some points on confusion matrix, I suggest that the authors do not introduce the paper by emphasizing the absence of ground truth on crowdsourced data sets.
    Section 4.1.2. The global confusion matrix should impose to aggregate the workers prediction, how should it be done (relatively to previous section)?
    Header of section 5 is a bit redundant with the introduction.
    End of 5.2.1, is there any recommandation on what to do with the "bad" workers? Is there any alternative to simple discard?
    On the example codes shown by the authors (that all ran on my laptop without problem), such as the ones for simulations, the displaying of graphs and tables is rather verbose as it is made "by hand" using pandas or matplotlib. I am not an expert in python, but would it be possible to include base plots or summaries for standard outputs of the main functions (such as the overloaded plot and summary function in R.) As the context of analyzing crowdsourced data might involve non-expert in python, I think that might be a useful feature.
    Overall, the arguments of the main functions (aggregate and aggregate-deep) might deserve a bit of explanation for the reader, has their name (in the current version of the code) is not really explicit (contrary to the arguments of the function simulate).

*Minor remarks*

    Page 2, item 3 "When the first question is..." should be "When the second question...", I guess;
    Section 2.1: Typo: "In praticular, given ..."
    The peerannot aggregate ./datasets/cifar10H/ -s GLAD line code launched the algorithm but after one minute waiting , the first iteration was not completed, is there a problem?"
    Section 4.1.1. In the formula of 

, shouldn't the role of and be switched, otherwise, this means we would compute, in the formula just above,
, which is 0 or 1, as far as I understand.
Section 4.1.2. I think
should be

    .

### Comments On Reproducibility:

From a reproductible point of view, the authors provide all working codes and all the results and figure depicted in the articles can be reproduced.

### Broader Impact Concerns:
I have no concern about any ethical problem. However, as said above, I am convinced that the authors could find a more relevant use case to show the potential interests of the crowdsourced data problematic.

>  **authors' answer:** 
>   Thank you for your detailed review. We appreciate your insight and respond below to your concerns:
> *Major remarks*
>    Dataset, ground truth and concrete use case: as this is your main concern, let us begin with that point. We want to clarify that we never use ground truth labels in our work during aggregation or training, as this is unknown information for practitioners. We have modified some of the pictures and part of the text to state that we only use an estimation of the true labels when needed. Yet, ground truth is possibly available for researchers in some datasets, for example for CIFAR-10H and LabelMe two of the most popular publicly available datasets in crowdsourcing classification (idem for the Music dataset, also available in peerannot). We agree that this is counter-intuitive, but comes from the way such datasets have been created. In such cases, the ground truth labels are available and can be used a posteriori for performance evaluation of the aggregation or agg-deep steps; see also recent works on these datasets (see [1], [2] and [3] for example from 2021/2022/2023). We also would like to emphasize that our goal, here, is not to create a new crowdsourcing dataset but to provide a toolbox to handle and reproduce existing experiments more easily. peerannot aggregate and identify commands can of course be run on a dataset with unknown ground truth. The identify one allowing you to detect bad workers for example is illustrated in Fig. 9 and 10 from section 5. We have also modified the structure of the dataset in Listing. 1 which implicitly required a ground truth for the directory containing the files. We have adapted the code accordingly to avoid such unfortunate dependency and are working on integrating more dataset structures in peerannot to fit users needs from different fields of crowdsourcing.
>    Data format: the format with -1 and the labels is actually a format proposed by Rodrigues et. al (2018) but not the one we recommand. In fact, depending on how the crowdsourcing was done (AmazonMTurk, Toloka, Appen, your own plateform, etc.) the format widely varies. We are working on a module in peerannot allowing to fit multiple plateforms format conversions. However, and this is a remark raised in this paper, the community could benefit on using a format that handles large sparse datasets like the .json we proposed, to possibly scale to large project (e.g., on Pl@ntnet data, the number of "workers" and taks could reach millions of units).
>    Comparison with other toolbox: We added an appendix to compare our framework with the crowd-kit library by Toloka. The main differences are that our identify module allows to detect bad workers/spammers with the spammer score by Raykar et. al (2017) or the trace confusion. We also provide the train command to directly learn (using Pytorch) from the aggregated labels. From labels aggregated with the strategy, this command allows users to load the dataset   and train a neural network specifying the optimizer, batch size and other classical hyperparameters. If a test set is available, predictive performance like test accuracy and calibration scores are computed. Crowd-kit does label aggregation on images classification and segmentation, we chose to focus on classification only. Finaly, we provide in Appendix a performance comparison of different aggregation strategies in term of computational time using the Benchopt [4] library on several datasets.

> Citations:

    [1]: H. Yang, X. Li and W. Pedrycz, "Learning From Crowds With Contrastive Representation," in IEEE Access, vol. 11, pp. 40182-40191, 2023, doi: 10.1109/ACCESS.2023.3269751.
    [2]: Zhendong, C., & Hongning, W. (2021). Improve Learning from Crowds via Generative Augmentation. Proceedings of the 27th ACM SIGKDDConference on Knowledge Discovery and Data Mining (KDD’21).
    [3]: Yang, W., Li, C. & Jiang, L. Learning from crowds with decision trees. Knowl Inf Syst 64, 2123–2140 (2022). https://doi.org/10.1007/s10115-022-01701-9
    [4]: Moreau, Thomas, et al. "Benchopt: Reproducible, efficient and collaborative optimization benchmarks." Advances in Neural Information Processing Systems 35 (2022): 25404-25421

## Second Round

Thank you for your replies to my comments.

They address all my points.

I have no further comments to make, and I approve publication of the article as it stands.
